db/write_thread.cc

```diff
@@ -558,18 +558,42 @@ void WriteThread::ExitAsBatchGroupLeader(WriteGroup& write_group,
     if (!leader->ShouldWriteToMemtable()) {
       CompleteLeader(write_group);
     }
+
+    // We have to link our group to memtable writer queue before wake up the
+    // next leader or set newest_writer_ to null, otherwise the next leader
+    // can run ahead of us and link to memtable writer queue before we do.
+    //
+    // However, last_writer can finish writing to memtable and join
+    // newest_writer_ queue again before we look for next leader. As a result,
+    // when we scan the queue and look for next leader, if we encounter
+    // last_writer, we won't be able to tell if we reach the current write
+    // group or not. Push a dummy writer to the queue so we can tell the two
+    // cases apart.
+    Writer dummy;
+    Writer* group_boundary = last_writer;
+    Writer* newest_writer = last_writer;
+    if (newest_writer_.compare_exchange_strong(newest_writer, &dummy)) {
```

Comment:
I guess "newest_writer_.compare_exchange_strong(last_writer, &dummy)" might be easier to read